src/views/Experiences/ExperiencesComponents/ExperiencesComponent.vue

```diff
@@ -42,15 +49,121 @@ const filteredExperiences = computed(() => {
+const isUpvote = (exid: number) => {
+  const vote = findVoteId(exid);
+  return vote && vote.helpfulness === 'up'; // Ensure it exists and is an upvote
+};
+
+const isDownvote = (exid: number) => {
+  const vote = findVoteId(exid);
+  return vote && vote.helpfulness === 'down'; // Ensure it exists and is a downvote
+};
+
+// Function to fetch all votes for a given experience ID
+const retrieveVote = async (exid: number) => {
+  if (auth.currentUser) {
+    try {
+      const response = await fetch(`${apiUrl}/experiences/${exid}/votes`);
+      if (!response.ok) {
+      throw new Error(`Response status: ${response.status}`);
+      }
+      votes.details = await response.json();
+      console.log(votes.details);
+    } catch (error) {
+      console.error(error);
+    }
+  } else {
+    console.log("User not logged in.")
+  }
+}
+
+// Handles the addition, editing, and deletion of votes
+const handleVote = async (exid:number, vote:string) => {
+  // If user is logged in and the vote does not exist, add vote
+  if (auth.currentUser && !checkExistingVote(exid) && vote) {
+    try {
+      await fetch(`${apiUrl}/experiences/${exid}/votes`, {
+      method: 'POST',
+      headers: {
+        'Content-Type': 'application/json',
+      },
+      body: JSON.stringify({
+        //vote data
+        userID: auth.currentUser.uid,
+        experienceID: exid,
+        vote: vote,
+      }),
+    });
+    } catch (error) {
+    console.error(error);
+    }
+    // if the user is logged in and the vote does exist...
+  } else if (auth.currentUser && checkExistingVote(exid) && vote) {
+    // find the vote to deal with
+    const voteToHandle = findVoteId(exid)
+    // if the vote info is the same, delete the vote
+    if (voteToHandle?.helpfulness === vote) {
+      try {
+        await fetch(`${apiUrl}/experiences/${exid}/votes/${voteToHandle?.vote_id}`, {
+          method: 'DELETE',
+          headers: {
+            'Content-Type': 'application/json',
+          },
+          body: JSON.stringify({
+            //vote data to delete
+            userID: auth.currentUser.uid,
+            experienceID: exid,
+            vote_id: voteToHandle?.vote_id
+          }),
+        });
+      } catch (error) {
+        console.error(error);
+      }
+      // if the vote info is not the same, edit it
+    } else {
+      try {
+        await fetch(`${apiUrl}/experiences/${exid}/votes/${voteToHandle?.vote_id}`, {
+          method: 'PATCH',
+          headers: {
+            'Content-Type': 'application/json',
+          },
+          body: JSON.stringify({
+            //vote data to edit
+            userID: auth.currentUser.uid,
+            experienceID: exid,
+            vote_id: voteToHandle?.vote_id
+          }),
+        });
+      } catch (error) {
+        console.error(error);
+      }
+    }
+  // other errors indicate user might not be logged in
+  } else {
+    console.log("User not logged in.")
+  }
+}
```

Comment:
Wow, this is very complex! I see now why you said "displeasure" on discord!

I'm a bit confused about why you want to delete a vote if the vote exists, but I'll trust your judgement on that since you designed this.

Reply:
The intention is so that lets say if someone already has an upvote on an experience, they can remove their vote if they wish by pressing on the upvote button again. I thought it would be better if users are able to revoke their vote as opposed needing to have an upvote or downvote on an experience after they casted it.